src/Symfony/Component/Form/ChoiceList/Factory/Cache/ChoiceLoader.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\ChoiceList\Factory\Cache;
+
+use Symfony\Component\Form\ChoiceList\Factory\CachingFactoryDecorator;
+use Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface;
+use Symfony\Component\Form\FormTypeInterface;
+
+/**
+ * A FlyWeight for types using static loaders, used by {@see CachingFactoryDecorator}.
+ *
+ * @internal
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+class ChoiceLoader implements ChoiceLoaderInterface
+{
+    private static $loaders = [];
+
+    /** @var ChoiceLoaderInterface */
+    private $loader;
+
+    /**
+     * @param FormTypeInterface     $formType A form type configuring a list
+     * @param ChoiceLoaderInterface $loader   A static loader used by the type
+     * @param mixed                 $vary     Dynamic data used to compute a unique hash when caching the loader
+     */
+    final public function __construct(FormTypeInterface $formType, ChoiceLoaderInterface $loader, $vary)
+    {
+        $uniquePrefix = $formType->getBlockPrefix();
+        $hash = $vary ? CachingFactoryDecorator::generateHash($vary) : '';
+
+        if (empty(self::$loaders[static::class][$uniquePrefix])) {
+            self::$loaders[static::class][$uniquePrefix][$hash] = $loader;
+        }
+
+        $this->loader = self::$loaders[static::class][$uniquePrefix][$hash];
+    }
+
+    final public function getLoader(): ChoiceLoaderInterface
+    {
+        return $this->loader;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoiceList($value = null)
+    {
+        return $this->loader->loadChoiceList($value);
+    }
+
+    /**
+     * Not implemented.
```

Comment:
`Not implemented.` — what does it mean?

Reply:
In my first draft I threw bad method call exceptions. This is just a flyweight, impelling the interface is being able to decorate the loader and pass the factory method because it **is** type hinted on the contrary to callables.
Not sure whether it should really implement those though. What do you and other think?

Reply:
It is safer to implement it in a way that we cannot know for sure that the caching factory decorator will be used, if a custom one is injected instead of decorating the existing. I may remove the comments and keep it as is I guess.

Reply:
I others agree, then I'll also reimplement `__invoke` in the abstract callback flyweight, for consistency, which I removed before decoupling this commit.

Reply:
Comments removed.